File: tests/native/test_plan_inspection.c
#include "support.h"
static char folder[F_PATH], repo[F_PATH], hydra[F_PATH], policy[F_PATH],
    estimates[F_PATH], origin[F_PATH];
static json_object *plan;
static json_object *cli(const char *op, const char *a, const char *b,
                        const char *c, const char *d, int status) {
  char *argv[] = {hydra,     "workflow", "plan",    (char *)op, (char *)a,
                  (char *)b, (char *)c,  (char *)d, NULL};
  struct f_capture r = nt_run(argv);
  json_object *v = nt_output(&r, status);
  f_capture_free(&r);
  return v;
}
static json_object *data(json_object *v) {
  json_object *out = nt_clone(f_field(v, "data"));
  json_object_put(v);
  return out;
}
static void compile(const char *name, json_object *p, char output[F_PATH],
                    char digest[65]) {
  char source[F_PATH], file[128];
  assert(snprintf(file, sizeof file, "%s.plan.json", name) > 0);
  nt_path(source, folder, file);
  nt_json_write(source, p ? p : plan);
  assert(snprintf(file, sizeof file, "%s.compiled.json", name) > 0);
  nt_path(output, folder, file);
  json_object_put(cli("compile", source, policy, output, NULL, 0));
  json_object *v = cli("show", output, "--json", NULL, NULL, 0);
  assert(!f_copy(digest, 65, f_string(f_field(v, "data"), "sha256")));
  json_object_put(v);
}
static void estimate(const char *id, int low, int high, json_object *plans) {
  json_object *e = f_parse("{\"source\":\"Explicit test "
                           "intervals\",\"cost_unit\":\"USD\",\"steps\":{}}");
  const char *names[] = {"spawn", "compose", "verify"};
  for (size_t i = 0; i < 3; i++) {
    json_object *s = json_object_new_object();
    const char *keys[] = {"milliseconds", "cost_microunits"};
    for (size_t j = 0; j < 2; j++) {
      json_object *a = json_object_new_array();
      json_object_array_add(a, json_object_new_int(low));
      json_object_array_add(a, json_object_new_int(high));
      json_object_object_add(s, keys[j], a);
    }
    json_object_object_add(f_field(e, "steps"), names[i], s);
  }
  json_object_object_add(plans, id, e);
}
static void write_estimates(const char *a, int low, int high, const char *b,
                            int blow, int bhigh) {
  json_object *v = f_parse("{\"schema_version\":1,\"plans\":{}}");
  estimate(a, low, high, f_field(v, "plans"));
  if (b)
    estimate(b, blow, bhigh, f_field(v, "plans"));
  nt_json_write(estimates, v);
  json_object_put(v);
}
static bool contains(json_object *array, const char *text) {
  for (size_t i = 0; i < json_object_array_length(array); i++)
    if (!strcmp(json_object_get_string(json_object_array_get_idx(array, i)),
                text))
      return true;
  return false;
}
static json_object *step(json_object *p, size_t i) {
  return json_object_array_get_idx(f_field(p, "steps"), i);
}
int main(void) {
  char tmp[] = "/tmp/hydra-plan-inspection-XXXXXX", source[F_PATH],
       home[F_PATH], left[F_PATH], right[F_PATH], lid[65], rid[65];
  assert(getcwd(origin, sizeof origin));
  nt_path(hydra, origin, "bin/hydra");
  nt_path(policy, origin, "tests/fixtures/plan/policy.json");
  nt_path(source, origin, "tests/fixtures/plan/plan.json");
  plan = f_read_json(source, 1000000);
  assert(plan);
  assert(mkdtemp(tmp));
  assert(!f_copy(folder, sizeof folder, tmp));
  nt_path(repo, folder, "repo");
  nt_path(source, origin, "tests/fixtures/plan/repo");
  nt_repo(source, repo);
  nt_path(home, folder, "home");
  nt_path(estimates, folder, "estimates.json");
  assert(!setenv("HYDRA_HOME", home, 1));
  if (!getenv("HYDRA_FLEET_BIN")) {
    nt_path(source, origin, "build/hydra-fleet");
    assert(!setenv("HYDRA_FLEET_BIN", source, 1));
  }
  assert(!chdir(repo));
  compile("base", NULL, left, lid);
  char *git[] = {"git", "status", "--porcelain", NULL};
  struct f_capture before = nt_run(git);
  json_object *v = data(cli("explain", left, NULL, NULL, NULL, 0)),
              *nodes = f_field(v, "nodes");
  assert(!strcmp(f_string(v, "plan_sha256"), lid));
  assert(json_object_array_length(nodes) == 3);
  const char *names[] = {"spawn", "compose", "verify"};
  for (size_t i = 0; i < 3; i++) {
    json_object *n = json_object_array_get_idx(nodes, i);
    assert(!strcmp(f_string(n, "id"), names[i]) &&
           !strcmp(f_string(n, "outcome_link"), "reachable"));
  }
  nt_equal(
      f_field(
          json_object_array_get_idx(
              f_field(json_object_array_get_idx(nodes, 1), "dependencies"), 0),
          "reasons"),
      "[\"creates_execution_head\"]");
  nt_equal(json_object_array_get_idx(
               f_field(json_object_array_get_idx(
                           f_field(json_object_array_get_idx(nodes, 2),
                                   "dependencies"),
                           0),
                       "reasons"),
               0),
           "{\"kind\":\"artifact_input\",\"input\":\"subject\",\"output\":"
           "\"report\"}");
  assert(!f_field(f_field(v, "metrics"), "critical_path_milliseconds") &&
         !f_field(f_field(v, "metrics"), "cost_microunits"));
  nt_equal(f_field(f_field(f_field(v, "metrics"), "hard_budgets"),
                   "timeout_seconds"),
           "180");
  json_object_put(v);
  struct f_capture after = nt_run(git);
  assert(!strcmp(before.out, after.out));
  f_capture_free(&before);
  f_capture_free(&after);
  compile("left", NULL, left, lid);
  json_object *changed = nt_clone(plan);
  json_object_object_add(f_field(step(changed, 1), "args"), "argv",
                         f_parse_value("[\"sh\",\"other-compose.sh\"]"));
  compile("right", changed, right, rid);
  json_object_put(changed);
  write_estimates(lid, 10, 20, rid, 100, 200);
  v = data(cli("compare", left, right, "--estimates", estimates, 0));
  nt_equal(f_field(v, "matched_scope"), "true");
  assert(!strcmp(f_string(v, "modeled_preference"), "left"));
  nt_equal(f_field(f_field(v, "left"), "critical_path_milliseconds"),
           "[30,60]");
  json_object *again =
      data(cli("compare", left, right, "--estimates", estimates, 0));
  assert(json_object_equal(v, again));
  json_object_put(again);
  nt_equal(f_field(json_object_array_get_idx(
                       f_field(f_field(v, "invalidation"), "checks"), 0),
                   "binding_changed"),
           "true");
  assert(
      strstr(f_string(f_field(v, "invalidation"), "reuse_policy"), "no reuse"));
  json_object_put(v);
  v = data(cli("compare", left, right, NULL, NULL, 0));
  assert(!strcmp(f_string(v, "modeled_preference"), "unresolved"));
  json_object_put(v);
  write_estimates(lid, 40000, 50000, rid, 100000, 200000);
  v = data(cli("compare", left, right, "--estimates", estimates, 0));
  nt_equal(f_field(f_field(v, "left"), "estimate_budget_conflict"), "true");
  assert(!strcmp(f_string(v, "modeled_preference"), "unresolved"));
  json_object_put(v);
  changed = nt_clone(plan);
  f_string_add(json_object_array_get_idx(f_field(changed, "requirements"), 0),
               "criterion", "A materially different target");
  compile("scope", changed, right, rid);
  json_object_put(changed);
  v = data(cli("compare", left, right, NULL, NULL, 0));
  nt_equal(f_field(v, "matched_scope"), "false");
  assert(contains(f_field(v, "scope_differences"), "requirements") &&
         !strcmp(f_string(v, "modeled_preference"), "unresolved"));
  json_object_put(v);
  const char *keys[] = {"envelope", "data"};
  for (size_t i = 0; i < 2; i++) {
    changed = nt_clone(plan);
    if (!i)
      json_object_object_add(f_field(changed, "envelope"), "timeout_seconds",
                             json_object_new_int(170));
    else
      json_object_object_add(
          f_field(f_field(f_field(f_field(f_field(changed, "data"), "steps"),
                                  "compose"),
                          "outputs"),
                  "report"),
          "max_bytes", json_object_new_int(1023));
    compile(keys[i], changed, right, rid);
    json_object_put(changed);
    v = data(cli("compare", left, right, NULL, NULL, 0));
    nt_equal(f_field(v, "matched_scope"), "false");
    assert(contains(f_field(v, "scope_differences"), keys[i]));
    json_object_put(v);
  }
  compile("bad-estimate", NULL, left, lid);
  write_estimates(lid, 1, 2, NULL, 0, 0);
  json_object *original = f_read_json(estimates, 1000000);
  const char *ranges[] = {"[-1,2]",  "[2,1]",    "[0,10000000000000]",
                          "[1.5,2]", "[true,2]", "[1]"};
  for (size_t i = 0; i < 8; i++) {
    changed = nt_clone(original);
    json_object *plans = f_field(changed, "plans"),
                *entry = f_field(plans, lid);
    if (i < 6)
      json_object_object_add(f_field(f_field(entry, "steps"), "spawn"),
                             "milliseconds", f_parse_value(ranges[i]));
    else if (i == 6) {
      json_object_object_add(
          plans,
          "0000000000000000000000000000000000000000000000000000000000000000",
          nt_clone(entry));
      json_object_object_del(plans, lid);
    } else
      json_object_object_add(f_field(entry, "steps"), "unknown",
                             json_object_new_object());
    nt_json_write(estimates, changed);
    json_object_put(changed);
    json_object_put(cli("explain", left, "--estimates", estimates, NULL, 1));
  }
  json_object_put(original);
  changed = f_read_json(left, 1000000);
  json_object_object_add(step(f_field(changed, "plan"), 0), "needs",
                         f_parse_value("[\"verify\"]"));
  nt_json_write(left, changed);
  json_object_put(changed);
  json_object_put(cli("explain", left, NULL, NULL, NULL, 1));
  changed = nt_clone(plan);
  json_object_array_add(f_field(step(changed, 2), "needs"),
                        json_object_new_string("spawn"));
  compile("partial", changed, left, lid);
  json_object_put(changed);
  write_estimates(lid, 5, 7, NULL, 0, 0);
  changed = f_read_json(estimates, 1000000);
  json_object_object_del(
      f_field(f_field(f_field(f_field(changed, "plans"), lid), "steps"),
              "compose"),
      "milliseconds");
  nt_json_write(estimates, changed);
  json_object_put(changed);
  v = data(cli("explain", left, "--estimates", estimates, NULL, 0));
  assert(!f_field(f_field(v, "metrics"), "critical_path_milliseconds"));
  nt_equal(f_field(f_field(v, "metrics"), "cost_microunits"), "[15,21]");
  nt_equal(f_field(f_field(v, "metrics"), "edges"), "3");
  json_object_put(v);
  json_object_put(plan);
  assert(!chdir(origin));
  nt_finish(folder, "Plan inspection: all six original case groups passed");
  return 0;
}

File: tests/test_plan_inspection.py
#!/usr/bin/env python3
"""Public, read-only plan explanations and matched comparisons; no execution."""
import copy
import json
import os
from pathlib import Path
import shutil
import subprocess
import tempfile
import unittest

ROOT = Path(__file__).resolve().parents[1]


class PlanInspection(unittest.TestCase):
    def setUp(self):
        self.temp = tempfile.TemporaryDirectory()
        self.addCleanup(self.temp.cleanup)
        self.path = Path(self.temp.name)
        self.repo = self.path / "repo"
        shutil.copytree(ROOT / "tests/fixtures/plan/repo", self.repo)
        self.env = {**os.environ, "HYDRA_HOME": str(self.path / "home"),
                    "HYDRA_FLEET_BIN": os.environ.get("HYDRA_FLEET_BIN", str(ROOT / "build/hydra-fleet"))}
        for args in (("init", "-q"), ("add", "."),
                     ("-c", "user.name=Test", "-c", "user.email=test@example.invalid", "commit", "-qm", "fixture")):
            subprocess.run(["git", *args], cwd=self.repo, check=True, capture_output=True)
        self.plan = json.loads((ROOT / "tests/fixtures/plan/plan.json").read_text())
        self.policy = ROOT / "tests/fixtures/plan/policy.json"

    def cli(self, *args, success=True):
        result = subprocess.run([str(ROOT / "bin/hydra"), "workflow", "plan", *map(str, args)],
                                cwd=self.repo, env=self.env, capture_output=True, text=True, timeout=30)
        self.assertEqual(result.returncode, 0 if success else 1, result.stdout + result.stderr)
        return json.loads(result.stdout)

    def compile(self, name, plan=None):
        source = self.path / f"{name}.plan.json"
        source.write_text(json.dumps(plan or self.plan))
        output = self.path / f"{name}.compiled.json"
        self.cli("compile", source, self.policy, output)
        digest = self.cli("show", output, "--json")["data"]["sha256"]
        return output, digest

    def estimates(self, *entries):
        data = {"schema_version": 1, "plans": {}}
        for digest, low, high in entries:
            data["plans"][digest] = {"source": "Explicit test intervals", "cost_unit": "USD",
                                      "steps": {name: {"milliseconds": [low, high], "cost_microunits": [low, high]}
                                                for name in ("spawn", "compose", "verify")}}
        file = self.path / "estimates.json"
        file.write_text(json.dumps(data))
        return file

    def test_explain_all_nodes_edges_and_unknowns(self):
        path, digest = self.compile("base")
        before = subprocess.check_output(["git", "status", "--porcelain"], cwd=self.repo)
        data = self.cli("explain", path)["data"]
        self.assertEqual(data["plan_sha256"], digest)
        self.assertEqual([n["id"] for n in data["nodes"]], ["spawn", "compose", "verify"])
        self.assertTrue(all(n["outcome_link"] == "reachable" for n in data["nodes"]))
        self.assertEqual(data["nodes"][1]["dependencies"][0]["reasons"], ["creates_execution_head"])
        reason = data["nodes"][2]["dependencies"][0]["reasons"][0]
        self.assertEqual(reason, {"kind": "artifact_input", "input": "subject", "output": "report"})
        self.assertIsNone(data["metrics"]["critical_path_milliseconds"])
        self.assertIsNone(data["metrics"]["cost_microunits"])
        self.assertEqual(data["metrics"]["hard_budgets"]["timeout_seconds"], 180)
        self.assertEqual(before, subprocess.check_output(["git", "status", "--porcelain"], cwd=self.repo))

    def test_ranges_are_bound_deterministic_and_never_authorize(self):
        left, lid = self.compile("left")
        changed = copy.deepcopy(self.plan)
        changed["steps"][1]["args"]["argv"] = ["sh", "other-compose.sh"]
        right, rid = self.compile("right", changed)
        file = self.estimates((lid, 10, 20), (rid, 100, 200))
        data = self.cli("compare", left, right, "--estimates", file)["data"]
        self.assertTrue(data["matched_scope"])
        self.assertEqual(data["modeled_preference"], "left")
        self.assertEqual(data["left"]["critical_path_milliseconds"], [30, 60])
        self.assertEqual(data, self.cli("compare", left, right, "--estimates", file)["data"])
        self.assertTrue(data["invalidation"]["checks"][0]["binding_changed"])
        self.assertIn("no reuse", data["invalidation"]["reuse_policy"])
        self.assertEqual(self.cli("compare", left, right)["data"]["modeled_preference"], "unresolved")
        file = self.estimates((lid, 40000, 50000), (rid, 100000, 200000))
        data = self.cli("compare", left, right, "--estimates", file)["data"]
        self.assertTrue(data["left"]["estimate_budget_conflict"])
        self.assertEqual(data["modeled_preference"], "unresolved")

    def test_scope_change_prevents_ranking(self):
        left, _ = self.compile("left")
        changed = copy.deepcopy(self.plan)
        changed["requirements"][0]["criterion"] = "A materially different target"
        right, _ = self.compile("right", changed)
        data = self.cli("compare", left, right)["data"]
        self.assertFalse(data["matched_scope"])
        self.assertIn("requirements", data["scope_differences"])
        self.assertEqual(data["modeled_preference"], "unresolved")

    def test_budget_or_output_contract_change_is_not_matched(self):
        left, _ = self.compile("left")
        for key in ("envelope", "data"):
            changed = copy.deepcopy(self.plan)
            if key == "envelope":
                changed[key]["timeout_seconds"] = 170
            else:
                changed[key]["steps"]["compose"]["outputs"]["report"]["max_bytes"] = 1023
            right, _ = self.compile(key, changed)
            data = self.cli("compare", left, right)["data"]
            self.assertFalse(data["matched_scope"])
            self.assertIn(key, data["scope_differences"])

    def test_malformed_stale_and_unknown_estimates_rejected(self):
        path, digest = self.compile("base")
        file = self.estimates((digest, 1, 2))
        original = json.loads(file.read_text())
        cases = []
        for values in ([-1, 2], [2, 1], [0, 10**13], [1.5, 2], [True, 2], [1]):
            data = copy.deepcopy(original)
            data["plans"][digest]["steps"]["spawn"]["milliseconds"] = values
            cases.append(data)
        data = copy.deepcopy(original)
        data["plans"]["0" * 64] = data["plans"].pop(digest)
        cases.append(data)
        data = copy.deepcopy(original)
        data["plans"][digest]["steps"]["unknown"] = {}
        cases.append(data)
        for data in cases:
            file.write_text(json.dumps(data))
            self.cli("explain", path, "--estimates", file, success=False)
        data = json.loads(path.read_text())
        data["plan"]["steps"][0]["needs"] = ["verify"]
        path.write_text(json.dumps(data))
        self.cli("explain", path, success=False)

    def test_partial_ranges_remain_unknown_and_extra_order_is_explicit(self):
        plan = copy.deepcopy(self.plan)
        plan["steps"][2]["needs"].append("spawn")
        path, digest = self.compile("base", plan)
        file = self.estimates((digest, 5, 7))
        estimates = json.loads(file.read_text())
        del estimates["plans"][digest]["steps"]["compose"]["milliseconds"]
        file.write_text(json.dumps(estimates))
        data = self.cli("explain", path, "--estimates", file)["data"]
        self.assertIsNone(data["metrics"]["critical_path_milliseconds"])
        self.assertEqual(data["metrics"]["cost_microunits"], [15, 21])
        self.assertEqual(data["metrics"]["edges"], 3)


if __name__ == "__main__":
    unittest.main()
